velox/functions/prestosql/ArrayFunctions.h

```diff
@@ -316,4 +320,89 @@ struct ArraySumFunction {
   }
 };
 
+/// Function Signature: array_duplicates(array(T)) -> array(T)
+/// where T must be coercible to bigint or varchar.
+/// Returns a set of elements that occur more than once in array.
+///
+/// Internally, we maintain a hashmap to indicate non-null tri-state:
+/// 1. key does not exist (key not in map),
+/// 2. unique (map-value=true),
+/// 3. duplicate (map-value=false).
+/// Meanwhile, maintain a std::optional to indicate null tri-state:
+/// 1. null does not exist (std::nullopt)
+/// 2. unique (value=true),
+/// 3. duplicate (value=false).
+template <typename TExecCtx, typename T>
+struct ArrayDuplicatesFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(TExecCtx);
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<velox::Array<T>>& out,
+      const arg_type<velox::Array<T>>& inputArray) {
+    folly::F14FastMap<arg_type<T>, bool> uniqueMap;
+    std::optional<bool> nullState = std::nullopt;
+    // TODO: Introducing an ordered set is redundant
+    //  if std::sort() in ArrayWriter is implemented
+    std::set<arg_type<T>> duplicateNonNullElements;
```

Comment:
I'm using `std::set` ([rb-tree internally](https://stackoverflow.com/a/11812871)) because I want all elements in an ascending order (that's the expected behavior of this function) when iterating it.
Per folly document, `F14FastSet` is a replacement of `std::unordered_set`. AFAIK, there is no equivalent replacement for `std::set` in folly family. There is a `F14VectorSet`, but it's in LIFO order without sorting.